kernel/kref.c

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sys/kref.h>
+
+void k_ref_init(struct k_ref *const ref)
+{
+	*ref = (struct k_ref){
+		.refs = ATOMIC_INIT(1),
+	};
+}
+
+struct k_ref *k_ref_get(struct k_ref *const ref)
+{
+	atomic_val_t old;
+
+	/* Reference counter must be checked to avoid incrementing ref from
+	 * zero, then we should return NULL instead.
+	 * Loop on compare-and-set in case someone has modified the reference
+	 * count since the read, and start over again when that happens.
```

Comment:
So here you're adding an extra few tests/branches and a second atomic operation (on top of what would otherwise be a simple atomic_inc()) to implement the return value that you documented earlier.  This seems like a poor trade?  Again, why do we want that?

Reply:
I looked at a few ref implementations like net_buf_ref and bt_conn_ref. This code is mostly copied from bt_conn_ref(which doesn't justify it or address your question). Given that we probably want to replace all ref-counts using this implementation I basically chose the one which was the most advanced.

Their intention is that they want to prevent increasing the reference count from 0 to 1. But I've also seen arguments within GitHub issues in this repo which say that passing an object with a count of 0 is basically the same as passing an invalid pointer so it'd probably be fine to just say "the behavior is undefined, we don't handle that".

Reply:
I removed that handling and added assertions instead.